controller/src/main/java/io/pravega/controller/task/Stream/StreamTransactionMetadataTasks.java

```diff
@@ -198,15 +207,20 @@ public Void initializeStreamWriters(final String commitStreamName, final EventSt
      * @param contextOpt optional context
      * @return true/false.
      */
-    @Task(name = "commitTransaction", version = "1.0", resource = "{scope}/{stream}/{txId}")
     public CompletableFuture<TxnStatus> commitTxn(final String scope, final String stream, final UUID txId,
                                                   final OperationContext contextOpt) {
-        final OperationContext context = getNonNullOperationContext(scope, stream, contextOpt);
+        return checkReady().thenComposeAsync(x -> {
+            final OperationContext context = getNonNullOperationContext(scope, stream, contextOpt);
+            return commitTxnBody(hostId, scope, stream, txId, context);
+        }, executor);
+    }
 
-        return execute(
-                new Resource(scope, stream, txId.toString()),
-                new Serializable[]{scope, stream, txId, null},
-                () -> commitTxnBody(scope, stream, txId, context));
+    private CompletableFuture<Void> checkReady() {
+        if (!ready) {
+            return FutureHelpers.failedFuture(new IllegalStateException(getClass().getName() + " not yet ready"));
```

Comment:
There are only two possible cases:
a) We never invoke any method on this class until initialization is completed, in which we could have just as easily performed the initialization later. (Then this is pointless complexity)
b) We have code that is getting exceptions thrown to it. (This would be even worse)
From your response it sounds like you're saying we're in case B, or at least we might be and have no way of telling. Is that correct?